### ISSUE TITLE
confusion about reproducing

### ISSUE THREAD

**Issue by vMagina** [lzx551402/contextdesc#19]
Now, I am trying to reproduce the contextdesc through pytorch.

- As described in the paper, during the training stage you sample 1,024 keypoint correspondences for each image pair. And this keypoint correspondences contain matchable keypoints as well as noisy keypoints. After I scaned the train dataset, I find that there isn't enough points in some image correspondence. So this images need to be remove during the training ? 

- Besides, as described in the dataset, the train dataset consist of undistorted images. So, the patches should generate from this undistorted images. Then, should do some scale transform berfore using the keypoint atrributes stored in the img_kpts.bin? 

- In the supplementary appendix, rewritten loss formulas are presented which cannot fully understand. So, when you train the model through the way : at first time, put matchable keypoints to the model; at second time, put the nosiy keypoints to the model. Or put all keypoints together?

- Last but not least, can you provide the whole train dataset list? I caculated the total image correspondences as mentioned in your Gl3D repository, the num is 1,442,252. So, you use all this pairs to train the model? And after how much steps, the loss is converged?

Can you answer my doubts？Thanks a lot.

**Comment by lzx551402** [comment]
1. Yes, we have a preprocessing step that discards the image pairs that have correspondences less than 16. For example, one pair contains 224 matches, and we then randomly sample another 800 points from the keypoint sets to form 1024 keypoints for training.

2. Yes, the input coordinates need to be undistorted first. Please refer to this [issue](https://github.com/lzx551402/GL3D/issues/6) that I just wrote an example to address.

3. Put all keypoints together.

4. I composed 800,000 image pairs for training. Given batch_size=2, it takes 400,000 iterations to complete one epoch - and we train only one epoch in our implementation.

**Comment by vMagina** [comment]
> 1. Yes, we have a preprocessing step that discards the image pairs that have correspondences less than 16. For example, one pair contains 224 matches, and we then randomly sample another 800 points from the keypoint sets to form 1024 keypoints for training.
> 2. Yes, the input coordinates need to be undistorted first. Please refer to this [issue](https://github.com/lzx551402/GL3D/issues/6) that I just wrote an example to address.
> 3. Put all keypoints together.
> 4. I composed 800,000 image pairs for training. Given batch_size=2, it takes 400,000 iterations to complete one epoch - and we train only one epoch in our implementation.

Thanks for your explanations.

**Comment by vMagina** [comment]
I encouter some problems during the training stage.

- My loss is huge. It is close to 4000 and it decay hardly. There is a big gap between this and the results in your paper. So, in the loss function, you only calculate the matchable keypoints? Or use the correspondence labels to make the product distance betweent nosiy keypoint pair slower? And this labels will be used in the hinge loss too?

- I use L2 regularization on the temperature parameter. But the value doesn't change. I think it is due to the huge loss. Also, I don't know why this parameter increasing constantly as showing in your paper. I think the L2 regularization should make it slower. Can you explain the phenomenon?
